feat/hiervis

kartoFlane opened the pull request:
Cleaned up version of previous pull request (#1)

...Forgot to replace temp name of the last commit with something meaningful. Extended description contains more information on what was actually changed.

kartoFlane committed:
Added documentation to interfaces.

kartoFlane committed:
Tweaks to comparator classes.

- StringIdComparator uses less comparisons now, added comments. Loop bound is computed once.
- NodeIdComparator now retains a single instance of StringIdComparator, used to perform all comparisons.

kartoFlane committed:
Added convenience constructor w/o trueClass param to BasicInstance

kartoFlane committed:
Added convenience constructors to BasicNode. Minor tweaks.

- Moved setNodeId() below constructors
- Code format changes.

kartoFlane committed:
Documentation for BasicNode.recalculateCentroid()

kartoFlane committed:
Improved BasicNode.printSubtree(). Added toString()

kartoFlane committed:
Added support for data column names in BasicHierarchy. Other changes.

- Added convenience constructors
- Added check to make sure root is never null
- Minor format changes, variable renames
- Added toString()

kartoFlane committed:
qwe

- Replaced HierarchyFiller with HierarchyBuilder
- Added withColumnHeaders argument to DataReader and its concrete implementations
- Added documentation to DataReader
- Reworked GeneratedCSVReader
--- Added loading of data column names, if the option is set
--- Fixed problems with loading files with UTF-8 encoding with uncommon characters
--- The load() method now throws an exception instead of exiting, allowing the client program to decide on the appropriate course of action.

lpolech commented on src/basic_hierarchy/interfaces/Instance.java:
```diff
@@ -1,9 +1,30 @@
 
-public interface Instance {
+/**
+ * An instance is a singular entity that can be placed in the broader context of a {@link Hierarchy}.
```
My proposition:
An instance is a singular entity (a single observation/data point) that can be placed in the {@link Node} that are placed in the broader context in a {@link Hierarchy}.

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
 	public int getNumberOfInstances();
+
+	/**
+	 * Returns the number of groups within the specified class.
```
Returns the number of instances that are assigned to {@code className} class within the whole hierarchy.

lpolech commented on src/basic_hierarchy/interfaces/Instance.java:
```diff
@@ -1,9 +1,30 @@
-	public String getNodeId();
+
+	/**
+	 * @return data columns associated with this instance
```
@return data row (feature values) associated with this instance

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
-	public int getClassCount(String className, boolean withClassHierarchy);
+
+	/**
+	 * Returns the total number of instances in this hierarchy, ie. sum of instances in all groups.
```
@return the total number of instances in this hierarchy, ie. sum of instances in all groups.

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
 	public String[] getClasses();
+
+	/**
+	 * Returns the number of ground-truth classes. Equivalent to {@code getClasses().length}.
```
@return the number of ground-truth classes. Equivalent to {@code getClasses().length}.

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
 	public int getNumberOfGroups();
+
+	/**
+	 * Returns an array of classes present in this hierarchy.
```
use @return

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
 	public Node[] getGroups();
+
+	/**
+	 * Returns the number of groups in this hierarchy. Equivalent to {@code getGroups().length}.
```
use @return

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
 	public Node getRoot();
+
+	/**
+	 * Returns an array of groups of instances.
```
use @return

lpolech commented on src/basic_hierarchy/interfaces/Hierarchy.java:
```diff
@@ -1,12 +1,54 @@
+public interface Hierarchy
+{
+	/**
+	 * Returns the root node of this hierarchy.
```
use @return

lpolech commented on src/basic_hierarchy/interfaces/Instance.java:
```diff
@@ -1,9 +1,30 @@
 	public double[] getData();
+
+	/**
+	 * @return id of the true class (group) to which this instance has been assigned ('ground truth').
```
@return id of the true class (group) from which this instance comes from ('ground truth').

lpolech commented on src/basic_hierarchy/interfaces/Instance.java:
```diff
@@ -1,9 +1,30 @@
+	 */
+	public String getNodeId();
+
+	public void setNodeId( String id );
```
missing javadoc?

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
-public interface Node {	
+
+/**
+ * A group is a collection of {@link Instance}s which belong to the same class.
```
A {@link Node} is a collection of {@link Instance}s which were assign to one {@link Node} by i.e. some method or generator.

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
+public interface Node
+{
+	/**
+	 * @return the id of this group.
```
@return the id of this node.

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
 
```
In this file you should replace "group" with "node" in the docs

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
-	 * Should returns instances belonging to this particular node
+
+	/**
+	 * Returns instances which belong to this particular group.
```
use @return

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
-	 * Should returns instances belonging to this node and its predecessors 
+
+	/**
+	 * Returns instances which belong to this group or its subgroups.
```
use @return

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
-	 * Could be node centroid or medoid
+
+	/**
+	 * Returns a node centroid or medoid.
```
@return node representation e.g. its centroid or medoid.

lpolech commented on src/basic_hierarchy/interfaces/Node.java:
```diff
@@ -2,36 +2,94 @@
+	 */
+	public void setRepresentation( Instance representation );
 
 	public void printSubtree();
```
javadoc?

lpolech reviewed (changes requested):
Find detailed comments in files

lpolech reviewed (approved):
I like that implementation!

lpolech reviewed (changes requested):
Why did you change nodeId to assignedClass? What is the purpose of this change? If so maybe it is worth to change class property from nodeId to assignedClass?

lpolech commented on src/basic_hierarchy/implementation/BasicNode.java:
```diff
@@ -2,114 +2,135 @@
-		for(Node child: children)
-		{
-			subtreeInstances.addAll(child.getSubtreeInstances());
+	public LinkedList<Instance> getSubtreeInstances()
```
What is the purpose of this change?

kartoFlane replied:
This is a simple variable rename, combined with code formatting changes made by autoformatter. Must've slipped through the purge of cosmetic changes.

lpolech replied:
Don't you like the name "subtreeInstances"?

kartoFlane replied:
I usually prefer to name the variable holding the return value as 'result', at least in simple functions.

lpolech commented on src/basic_hierarchy/implementation/BasicNode.java:
```diff
@@ -2,114 +2,135 @@
-    {
-        this.representation = representation;
-    }
+	public void setRepresentation( Instance representation )
```
I would prefer expand tab with 4 spaces

lpolech commented on src/basic_hierarchy/implementation/BasicHierarchy.java:
```diff
@@ -1,101 +1,183 @@
+		classCounts = new int[eachClassWithCount.size()];
+
+		LinkedList<String> sortedKeys = new LinkedList<String>( eachClassWithCount.keySet() );
+		sortedKeys.sort( new StringIdComparator() );
```
Unfortunately this code needs to support java 1.7.

lpolech commented on src/basic_hierarchy/reader/GeneratedCSVReader.java:
```diff
@@ -1,152 +1,195 @@
+public class GeneratedCSVReader implements DataReader
+{
+	/**
+	 * This method assumes that data are generated using Michał Spytkowski's data generator, using TSSB method.
```
What about adding a reference to the article that is describing this generator? https://arxiv.org/abs/1606.05681

kartoFlane replied:
Done.

lpolech commented on src/basic_hierarchy/reader/GeneratedCSVReader.java:
```diff
@@ -1,152 +1,195 @@
-     * in the first section to be the root of the tree. The nodes are given in depth first
-     * order."
-	 * @see interfaces.DataReader#load(java.lang.String)
+public class GeneratedCSVReader implements DataReader
```
What about splitting load function into smaller functions?

kartoFlane replied:
Should be possible. Will look into this later.

lpolech commented on src/basic_hierarchy/reader/GeneratedARFFReader.java:
```diff
@@ -19,8 +21,14 @@
 	@Override
-	public Hierarchy load(String filePath, boolean withInstancesNameAttribute, boolean withClassAttribute,
-						  boolean fillBreathGaps, boolean useSubtree) {
+	public Hierarchy load(
```
What about splitting load function into smaller functions? And maybe find some common abstractions to be used both in CSV and ARFF readers

lpolech commented on src/basic_hierarchy/common/HierarchyBuilder.java:
```diff
@@ -0,0 +1,395 @@
+			g.recalculateCentroid( useSubtree );
+		}
+
+		groups.sort( new NodeIdComparator() );
```
We need java 1.7 support ;(

kartoFlane replied:
Done. cb25b20

lpolech commented on src/basic_hierarchy/common/HierarchyBuilder.java:
```diff
@@ -0,0 +1,395 @@
+	 *            ID segments of the group as child
+	 * @return whether the two groups are in fact in a direct parent-child relationship.
+	 */
+	private static boolean areGroupsDirectlyRelated( String[] parentIds, String[] childIds )
```
I would suggest to change the name of this method to relate that we are comparing strings and checking if they are showing a parent->child relation

I don't like, when two methods have the same names with different parameters, because it is the sign that these two methods are doing different things.

kartoFlane replied:
Done. 57cd617

lpolech commented on src/basic_hierarchy/common/HierarchyBuilder.java:
```diff
@@ -0,0 +1,395 @@
+	 *            whether the cntroid calculation should also include child groups' instances.
+	 * @return the complete 'fixed' collection of groups, filled with artificial groups
+	 */
+	private static List<BasicNode> fixBreadthGaps( BasicNode root, List<BasicNode> groups, boolean useSubtree )
```
Can you split this method into smaller ones?

kartoFlane replied:
Done. 87b4dc5

lpolech commented on src/basic_hierarchy/common/HierarchyBuilder.java:
```diff
@@ -0,0 +1,395 @@
+	 * @return the complete 'fixed' collection of groups, filled with artificial groups
+	 */
+	private static List<BasicNode> fixDepthGaps( BasicNode root, List<BasicNode> groups, boolean useSubtree )
+	{
```
What about splitting this method into smaller ones?

kartoFlane replied:
Done. 5a3763d

lpolech commented on src/basic_hierarchy/common/HierarchyBuilder.java:
```diff
@@ -0,0 +1,395 @@
+			List<Node> newChildren = new LinkedList<Node>();
+
+			for ( int i = 0; i < children.size(); ++i ) {
+				Node childGroup = children.get( i );
```
Where do we assure that children will have null values when there will be no child node?

kartoFlane replied:
Now that I look at this, I feel like I might've made an incorrect assumption here. I'll check this later.

kartoFlane replied:
Indeed, this method was not working correctly. Fixed in ea3ad3b.

lpolech commented on src/basic_hierarchy/common/HierarchyBuilder.java:
```diff
@@ -0,0 +1,395 @@
+	 * @param fixBreadthGaps
+	 *            whether the hierarchy fixing algorithm should also fix gaps in breadth, not just depth.
+	 * @param useSubtree
+	 *            whether the centroid calculation should also include child groups' instances.
```
Add:
when set to {@code true} then all the objects from subnodes are also regarded as belonging to its supernodes.

kartoFlane committed:
Updated interface documentation

kartoFlane committed:
Changed 'group' -> 'node' for consistency

kartoFlane committed:
Changed argument name from 'assignedClass' to 'nodeId'

kartoFlane committed:
Reverted variable rename.

kartoFlane committed:
Changed sorting calls to use Collections.sort() instead.

kartoFlane committed:
Added note about the paper describing the data generator

kartoFlane committed:
Removed superfluous newlines in BasicNode.print()

kartoFlane committed:
Fixed fixBreadthGaps() not actually fixing any gaps.

kartoFlane committed:
Renamed id-comparing methods to be more clear. Updated javadocs.

- getNodeBranchIds() -> getNodeIdSegments()
- areNodesDirectlyRelated() -> areNodesParentAndChild() and areIdsParentAndChild()
- areNodesIndirectlyRelated() -> areNodesAncestorAndDescendant() and areIdsAncestorAndDescendant()

kartoFlane committed:
Extracted method for fixing breadth gaps in a single node from fixBreadthGaps()

kartoFlane committed:
Extracted method for fixing depth gaps between two specific nodes from fixDepthGaps(). Fixed a bug.

- Fixed a mistake which caused the algorithm to not consider artificial nodes (created during that method call) when looking for nearest ancestor.

kartoFlane committed:
Changed all gapfixing methods to return a collection of artificial nodes, not all nodes. Fixed javadoc refs.

kartoFlane committed:
Renamed HierarchyBuilder.buildHierarchy() to createParentChildRelations()

kartoFlane committed:
Added simple unit tests for fixDepthGaps() and fixBreadthGaps().

kartoFlane committed:
Reorganized CSVReader

- Fixed detection of access to file (was using && instead of ||)
- Extracted methods for improved readability (note: could still refactor the reader with a factory class)
- Added slightly better error / incorrect setting detection
- Added optional assertion that data rows in input file were sorted by assigned class column
- Changed whitespace from tabs to 4 spaces.
